src/System.Net.Http/tests/UnitTests/HPack/HPackRoundtripTests.cs

```diff
@@ -0,0 +1,180 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http.HPack;
+using System.Net.Http.Headers;
+using System.Text;
+using Xunit;
+
+namespace System.Net.Http.Unit.Tests.HPack
+{
+    public class HPackRoundtripTests
+    {
+
+        public static IEnumerable<object[]> TestHeaders =>
+            new HttpHeaders[]
+            {
+                BuildHttpHeaders(("header", new[] { "value" })),
+
+                BuildHttpHeaders(("header", new[] { "value1", "value2" })),
+
+                BuildHttpHeaders(
+                    ("header-0", new[] { "value1", "value2" }),
+                    ("header-0", new[] { "value3" }),
+                    ("header-1", new[] { "value1" }),
+                    ("header-2", new[] { "value1", "value2" }))
+
+            }.Select(h => new[] { h });
+
+        [Theory, MemberData(nameof(TestHeaders))]
+        public void HPack_Roundtrip_Headers(HttpHeaders headers)
```

Comment:
There's a lot going on in this test -- usually I'd expect unit tests to be very small. Is there any opportunity to simplify?

Reply:
I suppose it depends on what you mean by "a lot". Is the supporting test code too large, is it calling into too many components or both?

The test checks for one thing: a set of headers obtained after an encode/decode cycle should be structurally equal to the original input. As such I don't believe it's possible to simplify, other than perhaps introducing binary encoding constants and segregating encode and decode responsibilities into separate tests. I don't think this necessarily simplifies things though and moreover such kind of testing has already been established in the existing suites.

Reply:
Round-tripping is verified by functional tests already, so this could be changed to test only the non-public behavior (starting with a non-zeroed buffer).

If you choose to do this it would allow removing 80% of this logic like multiple headers, known headers, etc.

Reply:
Functional tests are slow though, so the number of values you can test is low. I think there's value in testing round-trips at the unit test level, a goal of mine is to eventually bring property-based testing and this would be a good candidate.

I agree that this imposes lots of duplication, however it's also an artifact of the asymmetry between the encoder and decoder implementation and significant of the encoder implementation being hardcoded in the Http2Connection class. I think we should eventually rework this (I presume we will, if we want to add support for dynamic tables), however I'm looking to get this fix ported to 3.0.